### ISSUE TITLE
Add support for AppSync pipeline resolvers

### ISSUE THREAD

**Issue by jrbeilke** [terraform-providers/terraform-provider-aws#8127]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Trying out the new AppSync resolvers that were added in 2.4.0 (#2705) but I'm not seeing a way to configure a pipeline resolver with multiple functions in the current implementation of `appsync_resolver`. Seems like the AWS Provider would need a new `appsync_function` resource and a modification to `appsync_resolver` allowing users to pass a pipeline_config.

https://docs.aws.amazon.com/appsync/latest/devguide/pipeline-resolvers.html

### New or Affected Resource(s)

* appsync_function
* appsync_resolver

### Potential Terraform Configuration

```hcl
resource "aws_appsync_graphql_api" "test" {
  authentication_type = "API_KEY"
  name                = "tf-example"
  schema              = <<EOF
type Mutation {
    putPost(id: ID!, title: String!): Post
}

type Post {
    id: ID!
    title: String!
}

type Query {
    singlePost(id: ID!): Post
}

schema {
    query: Query
    mutation: Mutation
}
EOF
}

resource "aws_appsync_datasource" "test" {
  api_id      = "${aws_appsync_graphql_api.test.id}"
  name        = "tf-example"
  type        = "HTTP"

  http_config {
    endpoint = "http://example.com"
  }
}

resource "aws_appsync_function" "test" {
  api_id      = "${aws_appsync_graphql_api.test.id}"
  data_source = "${aws_appsync_datasource.test.name}"
  name        = "tf-example"
  request_mapping_template = <<EOF
{
    "version": "2018-05-29",
    "method": "GET",
    "resourcePath": "/",
    "params":{
        "headers": $utils.http.copyheaders($ctx.request.headers)
    }
}
EOF
  response_mapping_template = <<EOF
#if($ctx.result.statusCode == 200)
    $ctx.result.body
#else
    $utils.appendError($ctx.result.body, $ctx.result.statusCode)
#end
EOF
}

resource "aws_appsync_resolver" "test" {
  api_id           = "${aws_appsync_graphql_api.test.id}"
  field            = "singlePost"
  type             = "Query"
  data_source      = "${aws_appsync_datasource.test.name}"
  request_template = <<EOF
{
    "version": "2018-05-29",
    "method": "GET",
    "resourcePath": "/",
    "params":{
        "headers": $utils.http.copyheaders($ctx.request.headers)
    }
}
EOF
  response_template = <<EOF
#if($ctx.result.statusCode == 200)
    $ctx.result.body
#else
    $utils.appendError($ctx.result.body, $ctx.result.statusCode)
#end
EOF

  pipeline_config  = {
    functions = ["${aws_appsync_function.test.name}"]
  }
}
```

### References

* #2705

**Comment by rimutaka** [comment]
@kterada0509 - thanks a lot for the pull request. It works for me without any apparent issues.

Can someone merge it into master, please?

**Comment by bflad** [comment]
Support for a new `aws_appsync_function` resource and `aws_appsync_resolver` arguments for `kind` and `pipeline_config` have been merged and will release with version 2.17.0 of the Terraform AWS Provider, likely later today. Thanks, @kterada0509 for the implementation!

**Comment by bflad** [comment]
This has been released in [version 2.17.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2170-june-28-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!
